File: app/src/main/java/com/stevenschoen/putionew/PutioApplication.java
package com.stevenschoen.putionew;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.support.multidex.MultiDexApplication;

import com.crashlytics.android.Crashlytics;
import com.stevenschoen.putionew.model.files.PutioFile;

import io.fabric.sdk.android.Fabric;

public class PutioApplication extends MultiDexApplication {
	private PutioUtils utils;

	@Override
	public void onCreate() {
		super.onCreate();

		Fabric.with(this, new Crashlytics());

        try {
            buildUtils();
        } catch (PutioUtils.NoTokenException e) {
            // User is not logged in
        }
	}

    public static PutioApplication get(Context context) {
        return (PutioApplication) context.getApplicationContext();
    }

    public boolean isLoggedIn() {
        if (utils != null) {
            SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);
            String token = sharedPrefs.getString("token", null);
            if (token != null && !token.isEmpty()) {
                return true;
            }
        } else {
            try {
                buildUtils();
                return isLoggedIn();
            } catch (PutioUtils.NoTokenException e) { }
        }

        return false;
    }

	public void buildUtils() throws PutioUtils.NoTokenException {
        utils = new PutioUtils(this);
    }

	public PutioUtils getPutioUtils() {
		return utils;
	}

    public interface CastCallbacks {
        void load(PutioFile file, String url, PutioUtils utils);
    }
}

File: app/src/main/java/com/stevenschoen/putionew/PutioApplication.kt
package com.stevenschoen.putionew

import android.app.Application
import android.arch.persistence.room.Room
import android.content.Context
import android.os.Build
import android.preference.PreferenceManager
import android.support.v4.app.Fragment
import com.stevenschoen.putionew.files.FileDownloadDatabase
import com.stevenschoen.putionew.model.files.PutioFile
import net.danlew.android.joda.JodaTimeAndroid
import timber.log.Timber

class PutioApplication : Application() {

    var putioUtils: PutioUtils? = null
        get() {
            if (field == null) {
                try {
                    field = PutioUtils(this)
                } catch (e: PutioUtils.NoTokenException) {
                    return null
                }
            }
            return field
        }
    val fileDownloadDatabase by lazy {
        Room.databaseBuilder(this, FileDownloadDatabase::class.java, "fileDownloads")
                .fallbackToDestructiveMigration()
                .build()
    }

    override fun onCreate() {
        super.onCreate()

        if (BuildConfig.DEBUG) {
            Timber.plant(Timber.DebugTree())
        } else {
            Timber.plant(CrashlyticsTree())
        }

        JodaTimeAndroid.init(this)

        if (Build.VERSION.SDK_INT >= 26) {
            createNotificationChannels(this)
        }
    }

    val isLoggedIn: Boolean
        get() {
            if (putioUtils == null) return false

            val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this)
            val token: String? = sharedPrefs.getString("token", null)
            return !token.isNullOrEmpty()
        }

    interface CastCallbacks {
        fun load(file: PutioFile, url: String, utils: PutioUtils)
        fun isCasting(): Boolean
    }
}

fun putioApp(context: Context) = context.applicationContext as PutioApplication
val Context.putioApp
    get() = putioApp(this)
val Fragment.putioApp
    get() = putioApp(context!!)
